### PR TITLE
Enable istio support for ray

### PR DESCRIPTION
**Pull request by hansinikarunarathne**
# Pull Request Template for Kubeflow manifests Issues

   I followed the steps in [https://docs.ray.io/en/latest/cluster/kubernetes/k8s-ecosystem/istio.html#step-4-create-the-raycluster](https://docs.ray.io/en/latest/cluster/kubernetes/k8s-ecosystem/istio.html#step-4-create-the-raycluster ) except the step 3. 
   
 
image
image
image
inside the notebook

## ✅ Contributor checklist
  - Make sure you have tested with kustomize. See [Installation Prerequisites](https://github.com/kubeflow/manifests#prerequisites)
  - All the commits have been _signed-off_  (To pass the `DCO` check)
  - Submit the [Contributor License Agreements](https://cla.developers.google.com/clas) (To pass the `cla/google` check)

---     
 
> You can join the CNCF Slack and access our meetings at the [Kubeflow Community](https://www.kubeflow.org/docs/about/community/) website. Our channel on the CNCF Slack is here [**#kubeflow-platform**](https://app.slack.com/client/T08PSQ7BQ/C073W572LA2).

### CODE REVIEW

File: contrib/ray/raycluster_example.yaml [hunk 1]
```diff
@@ -1,49 +1,29 @@
 apiVersion: ray.io/v1
 kind: RayCluster
 metadata:
-  name: kubeflow-raycluster
+  name: raycluster-istio
```
Comment (juliusvonkohout) [inline contrib/ray/raycluster_example.yaml:4]: Renaming this breaks the test. can you keep it as kubeflow-raycluster ?

Comment (juliusvonkohout) [reply contrib/ray/raycluster_example.yaml:4]: And we should install to the default user namespace if that is not already the case.

**Comment by juliusvonkohout** [comment]
@alex-treebeard @kevin85421 @wadhah101 can you take a look as well, since you contributed here before?

File: contrib/ray/raycluster_example.yaml [hunk 2]
```diff
@@ -2,48 +2,29 @@ apiVersion: ray.io/v1
 kind: RayCluster
 metadata:
   name: kubeflow-raycluster
+  namespace: kubeflow
```
Comment (juliusvonkohout) [inline contrib/ray/raycluster_example.yaml:5]: This is not the right namespace. It should be the customer/user namespace

Comment (hansinikarunarathne) [reply contrib/ray/raycluster_example.yaml:5]: Can you explain bit more

File: contrib/ray/raycluster_example.yaml [hunk 3]
```diff
@@ -57,7 +37,7 @@ spec:
           # entire Kubernetes node on which it is scheduled.
           resources:
             limits:
-              cpu: "1"
+              cpu: "2"
```
Comment (juliusvonkohout) [inline contrib/ray/raycluster_example.yaml:40]: Why did you change this?

File: contrib/ray/raycluster_example.yaml [hunk 4]
```diff
@@ -57,7 +37,7 @@ spec:
           # entire Kubernetes node on which it is scheduled.
```
Comment (juliusvonkohout) [inline contrib/ray/raycluster_example.yaml:37]: I think this comment can be deleted

File: contrib/ray/raycluster_example.yaml [hunk 5]
```diff
@@ -73,68 +53,52 @@ spec:
...
+            volumeMounts:
+             - mountPath: /tmp/ray
+               name: ray-logs
+            # The resource requests and limits in this config are too small for production!
+            # It is better to use a few large Ray pod than many small ones.
+            # For production, it is ideal to size each Ray pod to take up the
+            # entire Kubernetes node on which it is scheduled.
+            resources:
+             limits:
+               cpu: "2"
```
Comment (juliusvonkohout) [inline contrib/ray/raycluster_example.yaml:90]: Only change values if you have a valid reason

File: contrib/ray/raycluster_istio_auth_policy.yaml
Comment (juliusvonkohout) [inline contrib/ray/raycluster_istio_auth_policy.yaml:1]: This looks very dangerous.

File: .github/workflows/ray_test.yaml
Comment (juliusvonkohout) [inline .github/workflows/ray_test.yaml:1]: Where is the example namespace and multi-tenancy e.g. kubeflow-user-example-com from https://github.com/kubeflow/manifests/tree/master/common/user-namespace/base

File: contrib/ray/raycluster_istio_auth_policy.yaml
Comment (juliusvonkohout) [reply contrib/ray/raycluster_istio_auth_policy.yaml:1]: In the Kubeflow namespace there should be only kuberay operator, no raycluster worker or head. Raycluster,worker,head belong into the namespace/kubeflow profile kubeflow-user-example-com

File: contrib/ray/kuberay-operator/overlays/standalone/namespace.yaml
Comment (juliusvonkohout) [inline contrib/ray/kuberay-operator/overlays/standalone/namespace.yaml:1]: This file should be removed

File: contrib/ray/test.sh [hunk 6]
```diff
@@ -30,17 +30,27 @@ function trap_handler {
 
   # Delete KubeRay operator
   kustomize build kuberay-operator/base | kubectl -n $NAMESPACE delete -f -
+
 }
 
 trap trap_handler EXIT
 
+kubectl get namespaces --selector=istio-injection=enabled
```
Comment (juliusvonkohout) [inline contrib/ray/test.sh:38]: kubeflow-user-example-com is the namespace. It must be a proper multi-tenancy supported kubeflow profile namespace

File: contrib/ray/raycluster_example.yaml (hunk 5 above)
Comment (juliusvonkohout) [reply contrib/ray/raycluster_example.yaml:90]: We will also get a cleaner difference in git if you only change relevant values.

File: contrib/ray/raycluster_istio_auth_policy.yaml
Comment (juliusvonkohout) [reply contrib/ray/raycluster_istio_auth_policy.yaml:1]: The authorization policy should be in the file contrib/ray/raycluster_example.yaml as well as additional resource and it should only allow the istio service mesh traffic between ray workers, not more and not less.

**Comment by hansinikarunarathne** [comment]
did the requested changes. please review @juliusvonkohout

File: contrib/ray/raycluster_example.yaml [hunk 7]
```diff
@@ -1,60 +1,112 @@
...
+kind: AuthorizationPolicy
+metadata:
+  name: allow-ray-workers-head
+spec:
+  action: ALLOW
+  rules:
+  - to:
+    - operation:
+        ports:
+        - "6379"
```
Comment (juliusvonkohout) [inline contrib/ray/raycluster_example.yaml:11]: Please make sure that you only allow traffic from the default-editor serviceaccount (the one that the workers use as well) in the same namespace

File: contrib/ray/raycluster_example.yaml [hunk 8]
```diff
@@ -1,60 +1,112 @@
...
+    appProtocol: grpc
+  - name: p10010
+    port: 10010
+    appProtocol: grpc
+  - name: p10011
+    port: 10011
+    appProtocol: grpc
+  - name: p10012
+    port: 10012
+    appProtocol: grpc
```
Comment (juliusvonkohout) [inline contrib/ray/raycluster_example.yaml:75]: Why do you need so many ports and why is only one in the authorizationpolicy?

File: contrib/ray/raycluster_example.yaml [hunk 9]
```diff
@@ -1,60 +1,112 @@
...
+      max-worker-port: '10012'
+      node-ip-address: $(hostname -I | tr -d ' ' | sed 's/\./-/g').raycluster-istio-headless-svc.kubeflow-user-example-com.svc.cluster.local
     template:
-      metadata:
-        # Custom labels. NOTE: To avoid conflicts with KubeRay operator, do not define custom labels start with `raycluster`.
-        # Refer to https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/
-        # The ray head must not have an Istio sidecar
-        # TODO add an authorizationpolicy in the future for the ray head
-        labels:
-          sidecar.istio.io/inject: "false"
```
Comment (juliusvonkohout) [inline contrib/ray/raycluster_example.yaml:35]: can you keep 

```
        labels:
          sidecar.istio.io/inject: "true"
```
to have it explicit ?

File: contrib/ray/test.sh [hunk 10]
```diff
@@ -2,46 +2,30 @@
...
-    fi
-    sleep $SLEEP_INTERVAL
-  done
-
-  # Delete KubeRay operator
-  kustomize build kuberay-operator/base | kubectl -n $NAMESPACE delete -f -
-}
-
-trap trap_handler EXIT
+while [[ $(kubectl get namespace $NAMESPACE --no-headers 2>/dev/null | wc -l) -eq 0 ]]; do
```
Comment (juliusvonkohout) [inline contrib/ray/test.sh:10]: No while loops, this can run indefinitely. Please use a for loop with a maximum time

File: contrib/ray/raycluster_example.yaml (hunk 9 above)
Comment (hansinikarunarathne) [reply contrib/ray/raycluster_example.yaml]: Did it

File: contrib/ray/raycluster_example.yaml (hunk 8 above)
Comment (hansinikarunarathne) [reply contrib/ray/raycluster_example.yaml]: --node-manager-port: Raylet port for node manager.
--object-manager-port: Raylet port for object manager.
--runtime-env-agent-port: Raylet port for runtime env agent.
--dashboard-agent-grpc-port: The port to listen for grpc on
--max-worker-port: Maximum port number worker can be bound to.
--metrics-export-port: The port to use to expose Ray metrics.

These ports are needed when using the Ray Cluster. The access is given for the relevant ports through the authorization policy

File: contrib/ray/raycluster_example.yaml [hunk 11]
```diff
@@ -1,60 +1,96 @@
...
+kind: AuthorizationPolicy
+metadata:
+  name: allow-ray-workers-head
+spec:
+  action: ALLOW
+  rules:
+  - from:
+    - source:
+        principals:
+        - "cluster.local/ns/kubeflow-user-example-com/sa/kubeflow-raycluster"
```
Comment (juliusvonkohout) [inline contrib/ray/raycluster_example.yaml:11]: This must be generic. The namespace could also have another name that you do not know in advance. The same manifests should work for different users.

File: contrib/ray/test.sh [hunk 12]
```diff
@@ -2,46 +2,44 @@
...
+  if [ "$elapsed_time" -ge "$TIMEOUT" ]; then
+    echo "Timeout exceeded. Namespace $NAMESPACE not created."
+    exit 1
+  fi
+  
+  echo "Waiting for namespace $NAMESPACE to be created..."
+  sleep 2
+done
+
+echo "Namespace $NAMESPACE is created!"
```
Comment (juliusvonkohout) [inline contrib/ray/test.sh:29]: Has been created

File: contrib/ray/raycluster_example.yaml
Comment (juliusvonkohout) [inline contrib/ray/raycluster_example.yaml:1]: Please post a screenshot of the dashboard

File: contrib/ray/raycluster_example.yaml (hunk 11 above)
Comment (hansinikarunarathne) [reply contrib/ray/raycluster_example.yaml:11]: Is there a way to templating that namespace value without using helm?

Comment (juliusvonkohout) [reply contrib/ray/raycluster_example.yaml:11]: The Serviceaccount is wrong. It should be default-editor everywhere. Especially in the raycluster itself. 

You can take a look at https://istio.io/latest/docs/reference/config/security/authorization-policy/ for ideas and ask gpt4. Maybe you can just whitelist the whole namespace.

File: contrib/ray/raycluster_example.yaml
Comment (hansinikarunarathne) [reply contrib/ray/raycluster_example.yaml:1]: image
image
image
inside the notebook

File: contrib/ray/raycluster_example.yaml (hunk 11 above)
Comment (hansinikarunarathne) [reply contrib/ray/raycluster_example.yaml:11]: Is it only enough to generalize the service account name(with default-editor)? Because since we are deploying it in a namespace, we have to specify the namespace to give the access to only for that specific namespace(kubeflow-user-example in this case)

File: contrib/ray/raycluster_example.yaml
Comment (hansinikarunarathne) [reply contrib/ray/raycluster_example.yaml:1]: I added ss to descriptions as well

File: contrib/ray/test.sh (hunk 12 above)
Comment (hansinikarunarathne) [reply contrib/ray/test.sh:29]: Resolved

File: contrib/ray/raycluster_example.yaml (hunk 11 above)
Comment (juliusvonkohout) [reply contrib/ray/raycluster_example.yaml]: No the other way around. We need to generalize the namespace. It would be alright to allow all serviceaccounts or all pods from the namespace. But the deployments should still run as default-editor. This must be deployable without manual changes per namespace.

Comment (juliusvonkohout) [reply contrib/ray/raycluster_example.yaml]: Are you setting the default-editor in the raycluster resource itself?

File: contrib/ray/raycluster_example.yaml [hunk 13]
```diff
@@ -73,68 +109,51 @@ spec:
...
+            securityContext:
+             allowPrivilegeEscalation: false
+             capabilities:
+               drop: ["ALL"]
+             runAsNonRoot: true
+             seccompProfile:
+               type: RuntimeDefault
+          volumes:
+            - name: ray-logs
+              emptyDir: {}
```
Comment (juliusvonkohout) [inline contrib/ray/raycluster_example.yaml:159]: I am missing the Serviceaccount default-editor

File: contrib/ray/raycluster_example.yaml (hunk 11 above)
Comment (hansinikarunarathne) [reply contrib/ray/raycluster_example.yaml]: I added the default-editor service account to the deployment. 
> We need to generalize the namespace.

This means should I add like this,
cluster.local/ns/kubeflow-user-example-com/sa/*

File: contrib/ray/raycluster_example.yaml (hunk 13 above)
Comment (hansinikarunarathne) [reply contrib/ray/raycluster_example.yaml:161]: added

File: contrib/ray/raycluster_example.yaml (hunk 11 above)
Comment (hansinikarunarathne) [reply contrib/ray/raycluster_example.yaml]: I did the changes . @juliusvonkohout

**Comment by juliusvonkohout** [comment]
Please remove the part of "## Step 3: Create a namespace" there shall be no manual namespace creation. All of this is handled by kubeflow profiles.
